### PR TITLE
Rewrite the path close section to specify symmetric Abandon

### PR DESCRIPTION
**Pull request by huitema**
As discussed during the interim. This solves issue #313.

### CODE REVIEW

File: draft-ietf-quic-multipath.md [hunk 1]
```diff
@@ -537,25 +538,31 @@ can send ack-eliciting packets such as packets containing PING frames
...
+MUST do so explicitly by sending a PATH_ABANDON frame, as defined in
+{{path-close}}.
+
+### Early Abandon {#abandon-early}
+
+The are are scenarios in which an endpoint will receive a PATH_ABANDON frame
+before receiving or sending any traffic on a path. For example, if the client
+tries to initiate a path and the path cannot be establish, it will send a
+PATH_ABANDON frame (see {{path-initiation}}). An endpoint may also decide
+to abandon a path for any reason, such as for example removing a hole from
```
Comment (Yanmei-Liu) [inline draft-ietf-quic-multipath.md:556]:
```suggestion
to abandon a path for any reason, for example, removing a hole from
```

**Review by Yanmei-Liu — COMMENTED** [review]
Thanks a lot for this PR! Some minor fixes.

**Comment by huitema** [comment]
@Yanmei-Liu thanks for the review and the fixes!

**Review by Yanmei-Liu — APPROVED** [review]

File: draft-ietf-quic-multipath.md [hunk 2]
```diff
@@ -515,6 +480,42 @@ the server MAY wait for a short, limited time such as one PTO if a path
...
+### Handling ACK_MP for abandoned paths {#ack-after-abandon}
+
+When an endpoint decides to send a PATH_ABANDON frame, there may
+still be some unacknowledged packets. Some other packets may well
+be in transit, and could be received shortly after sending the
+PATH_ABANDON frame. As specified above, the endpoints SHOULD
+send ACK_MP frames promptly, to avoid unnecessary data
+retransmission after the peer deletes path resources.
+
+These ACK_MP frames SHOULD be send on a different path than the
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:513]: I think this is a MUST. Or no normatively language here at all. I would rather say somewhere else that if the PATH_ABANDON frame was sent, the endpoint MUST NOT send on that path anymore.

Comment (huitema) [reply draft-ietf-quic-multipath.md:513]: Yes, that's probably the best way to handle it, but there is the slight issue of abandoning the last path.

File: draft-ietf-quic-multipath.md [hunk 3]
```diff
@@ -438,62 +437,29 @@ connectivity or local preferences. After an endpoint abandons
...
+(see {(idle-time-close}}) or for any other reason, such as management
+of local resources. It is also possible to abandon a path for which no
+packet has been sent (see {{abandon-early}}).
+
+When an endpoint receives a PATH_ABANDON frame, it MUST send a corresponding
+PATH_ABANDON frame if it has not already done so, and it MUST treat all
+connection identifiers received from the peer for that path as immediately
+retired. However, the endpoint should retain some knowledge of the
+connection identifiers sent to the peer for a short period,
+as specified in {{spurious-stateless-reset}}.
```
[the remark this answers is not in the document]
Comment (huitema) [reply draft-ietf-quic-multipath.md:456]: I tried to develop the ACK_MP issue in the subsection "Handling ACK_MP for abandoned paths {#ack-after-abandon}". Maybe we should call that out here, in the same way that we call out stateless reset.

[the remark this answers is not in the document]
Comment (huitema) [reply draft-ietf-quic-multipath.md:456]: I also think that we have a hard rule on the data sender side: "if a path info is deleted, then all pending packets MUST be considered lost", and softer rule "don't delete the info too soon because you will end up with spurious retransmit". On the data receiver side, we could have a hard rule, "you MUST send an ACK_MP for all packets received before abandon by the peer", and a softer rule "you should wait a bit for packets to arrive after abandon so you can ACK them and avoid forcing the peer to resend them". Maybe develop that a bit more in the dedicated subsection.

File: draft-ietf-quic-multipath.md (hunk 2 above)
Comment (mirjak) [reply draft-ietf-quic-multipath.md:513]: For the last path you should do connection close instead. That’s already written somewhere but it might be a SHOULD only. If so, we can maybe change to a MUST.

Comment (huitema) [reply draft-ietf-quic-multipath.md:513]: Not sure it matters. Suppose that we treat "abandon of last path" as a protocol error. That means the peer receiving the abandon will automatically issue "connection close" with reason protocol error, and the peer sending the abandon will send nothing, because it must not send data after abandon. But yes, sending connection close is more graceful.

File: draft-ietf-quic-multipath.md [hunk 4]
```diff
@@ -515,6 +480,42 @@ the server MAY wait for a short, limited time such as one PTO if a path
...
+connection and could trigger the peer to send a Stateless Reset
+packet. The rule to "retain knowledge of connection ID for a short period"
+is intended to reduce the risk of sending such spurious stateless
+packets, but it cannot completely avoid that risk.
+
+The immediate retirement of connection identifiers received for the
+path guarantees that spurious stateless reset packets
+sent by the peer will not cause the closure of the QUIC connection.
+
+### Handling MP_ACK for abandoned paths {#ack-after-abandon}
```
[the remark this answers is not in the document]
Comment (huitema) [reply draft-ietf-quic-multipath.md:511]: Sorry, updates lost in a github snafu. Resetting.

File: draft-ietf-quic-multipath.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (huitema) [reply draft-ietf-quic-multipath.md]: @mirjak I think that the "tighten" commit fixes this. Please review.

File: draft-ietf-quic-multipath.md (hunk 2 above)
Comment (huitema) [reply draft-ietf-quic-multipath.md]: Replace SHOULD by MUST.

File: draft-ietf-quic-multipath.md [hunk 5]
```diff
@@ -361,11 +361,10 @@ unless it has previously validated that address.
 
 If validation succeeds, the client can continue to use the path.
 If validation fails, the client MUST NOT use the path and can
-remove any status associated to the path initation attempt.
+remove any status associated to the path initiation attempt.
 However, as the used Path ID is anyway consumed,
-and the endpoint MUST abandon the path by sending a PATH_ABANDON frame
-and retire all corresponding connection IDs by sending MP_RETIRE_CONNECTION_ID frames
-on another path to inform the peer that the Path ID cannot be used anymore.
+and the endpoint MUST explicitly close the path as specified in
```
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:366]:
```suggestion
the endpoint MUST explicitly close the path as specified in
```

File: draft-ietf-quic-multipath.md [hunk 6]
```diff
@@ -515,6 +486,43 @@ the server MAY wait for a short, limited time such as one PTO if a path
...
+The peers that send a PATH_ABANDON frame MUST treat all connection
+identifiers received from the peer for the path ID as immediately
+retired. The Stateless Reset Tokens associated with these connection
+identifiers MUST NOT be used to identify Stateless Reset packets
+per {{Section 10.3 of QUIC-TRANSPORT}}.
+
+Due to packet losses and network delays, packets sent on the path may
+well arrive after the PATH_ABANDON frames have been sent or received.
+If these packets arrive after the connection identifiers sent to the peer
+have been retired, they will not be recognizes as bound for the local
```
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:500]:
```suggestion
have been retired, they will not be recognized as bound for the local
```

**Review by qdeconinck — APPROVED** [review]
Two typos, but looks good to me

File: draft-ietf-quic-multipath.md [hunk 7]
```diff
@@ -558,25 +568,31 @@ can send ack-eliciting packets such as packets containing PING frames
...
-This means if all paths remain idle for the idle timeout, the connection
-is implicitly closed.
-
 Server implementations need to select the sub-path idle timeout as a trade-
 off between keeping resources, such as connection IDs, in use
 for an excessive time or having to promptly re-establish a path
 after a spurious estimate of path abandonment by the client.
 
+Endpoints that desire to close a path because of the idle timer rule
+MUST do so explicitly by sending a PATH_ABANDON frame, as defined in
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:577]:
```suggestion
MUST do so explicitly by sending a PATH_ABANDON frame on another active path, as defined in
```

File: draft-ietf-quic-multipath.md [hunk 8]
```diff
@@ -645,75 +661,8 @@ The peer that sends the MP_RETIRE_CONNECTION_ID frame can keep sending data
...
-
-"Closing" state is entered after an PATH_ABANDON frame was sent or received.
-In this state, the endpoint waits for three PTOs before sending MP_RETIRE_CONNECTION_ID frames.
-This allows for graceful tear down and processing of in-flight packets.
-
-When a path reaches the "Closed" state, the endpoint releases all the
-path's associated resources, including the associated connection IDs and the path identifier.
 
-# Packet Protection {#multipath-aead}
+## Packet Protection {#multipath-aead}
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:665]:
```suggestion
# Packet Protection {#multipath-aead}
```

File: draft-ietf-quic-multipath.md (hunk 7 above)
Comment (Yanmei-Liu) [reply draft-ietf-quic-multipath.md:577]: I think it's better to allow endpoints to send PATH_ABANDON frame either on the same path or another active path. This depends on whether the same path is a blackhole. So we don't need to specify here.

File: draft-ietf-quic-multipath.md [hunk 9]
```diff
@@ -461,60 +460,35 @@ connectivity or local preferences. After an endpoint abandons
...
-Reception or sending of the PATH_ABANDON frame is
-the first step to release all resources related to a
-Path ID. However, the Path ID can only be released after all active
-connection IDs for the Path ID have been retired or timed-out after
-the PATH_ABANDON frame was sent.
-Still, when an endpoint receives an PATH_ABANDON frame,
-it SHOULD NOT use the associated Path ID in future frames, except
-in ACK_MP frames for acknowledging inflight packets and
-in MP_RETIRE_CONNECTION_ID frames for connection ID retirement.
+does not indicate path abandonment (see {{retire-cid-close}}).
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:467]:
```suggestion
does not indicate path abandonment (see further {{consume-retire-cid}}).
```

File: draft-ietf-quic-multipath.md [hunk 10]
```diff
@@ -932,7 +881,7 @@ for example the PTO timeout: if an ACK_MP is not received after more
 than 350ms, either the data packet or its ACK_MP were probably lost.
 
 The simplest implementation is to compute smoothedRTT and RTTvar per
-{{Section 5.3 of QUIC-RECOVERY}} regardless of the path through which MP_ACKs are
+{{Section 5.3 of QUIC-RECOVERY}} regardless of the path through which ACK_MPs are
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:884]:
```suggestion
{{Section 5.3 of QUIC-RECOVERY}} regardless of the path through which ACK_MP frames are
```

File: draft-ietf-quic-multipath.md [hunk 11]
```diff
@@ -461,60 +460,35 @@ connectivity or local preferences. After an endpoint abandons
...
 connection ID retirement, the inverse is not true: retiring the associated connection IDs
 does not indicate path abandonment (see further {{consume-retire-cid}}).
-Implicit signals such as idle time or packet losses might be
-the only way for an endpoint to detect path closure (see {(idle-time-close}})
-if connectivity is broken on that path.
-
-Note that other explicit closing mechanisms of {{QUIC-TRANSPORT}} still
-apply on the whole connection. In particular, the reception of either a
-CONNECTION_CLOSE ({{Section 10.2 of QUIC-TRANSPORT}}) or a Stateless
-Reset ({{Section 10.3 of QUIC-TRANSPORT}}) closes the connection.
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:476]: @huitema why did you remove this paragraph?

Comment (huitema) [reply draft-ietf-quic-multipath.md:476]: Can't remember why. Restore it if you want.
